### PR TITLE
feat(2346): Support for job level parameters

### PR DESCRIPTION
**Pull request by sagar1312**
## Context

Ability to define parameters at job level

## Objective

Create builds with job level parameters

## References

[screwdriver-cd/screwdriver#2346](https://github.com/screwdriver-cd/screwdriver/issues/2346)

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: app/components/pipeline-event-row/component.js [hunk 1]
```diff
@@ -31,13 +31,47 @@ export default Component.extend({
...
+
+    const pipelineParameters = {};
+    const jobParameters = {};
+
+    // Segregate pipeline level and job level parameters
+    Object.entries(this.getWithDefault('event.meta.parameters', {})).forEach(
+      ([propertyName, propertyVal]) => {
+        const keys = Object.keys(propertyVal);
+
+        if (keys.length === 1 && keys[0] === 'value') {
```
Comment (adong) [inline app/components/pipeline-event-row/component.js:45]: // if (keys.length === 1 && keys[0] === 'value') { 
 
I don't follow the logics here. It's not obvious to me how to distinguish between job level and pipeline level parameters.

File: app/components/pipeline-event-row/component.js [hunk 2]
```diff
@@ -61,7 +95,7 @@ export default Component.extend({
     }
   }),
 
-  isCommiterDifferent: computed(
+  isCommitterDifferent: computed(
```
Comment (adong) [inline app/components/pipeline-event-row/component.js:98]: Thank you.

**Comment by adong** [comment]
Looks like validator works fine.
image

**Comment by adong** [comment]
For brand new pipelines, only shared parameters shows up, job level parameters does not show.
image

It shows after kick off the 1st build.

File: app/components/pipeline-event-row/component.js (hunk 1 above)
Comment (adong) [reply app/components/pipeline-event-row/component.js:45]: image

image

So for parameters, if its pipeline level, then key will be the pipeline level key, and value will be an object with single object such as 

```
{
    "main": {    // job name
        "p1": {    // job level
            "value": "main p1"
        },
        "p2": {  // job level
            "value": "parameter name for main job"
        }
    },
    "p1": {     // pipeline level
        "value": "p1"
    },
    "p2": {   // pipeline level
        "value": "adong123123"
    }
}
```

if it's job level parameters, the job name will be the key, and the values will be a nested object.

**Comment by adong** [comment]
1) Add a job name also names p1
```
shared:
    image: node:14

parameters:
  p1: p1
  p2:
      value: adong
      description: User running build

jobs:
    main:
        requires: [~pr, ~commit]
        steps:
            - p1: echo $(meta get parameters.p1)
            - p2: echo $(meta get parameters.p2)

        parameters:
            p1: main p1
            p2:
                value: parameter name for main job
                description: This is a job parameter desc
    p1:
        requires: [~pr, ~commit]
        steps:
            - p1: echo $(meta get parameters.p1)
            - p2: echo $(meta get parameters.p2)
        parameters:
            p1: p1's p1
```

image

2) Remove p1 from pipeline's parameters
```
shared:
    image: node:14

parameters:
  p2:
      value: adong
      description: User running build

jobs:
    main:
        requires: [~pr, ~commit]
        steps:
            - p1: echo $(meta get parameters.p1)
            - p2: echo $(meta get parameters.p2)

        parameters:
            p1: main p1
            p2:
                value: parameter name for main job
                description: This is a job parameter desc
    p1:
        requires: [~pr, ~commit]
        steps:
            - p1: echo $(meta get parameters.p1)
            - p2: echo $(meta get parameters.p2)
        parameters:
            p1: p1's p1
```
image

I tired few cases, and they look good to me.

**Review by adong — APPROVED** [review]

File: app/components/pipeline-event-row/component.js [hunk 3]
```diff
@@ -31,13 +31,47 @@ export default Component.extend({
...
+      pipelineParameters,
+      jobParameters
+    });
+  },
 
-      return Object.keys(parameters).length;
+  numberOfParameters: computed(
+    'pipelineParameters',
+    'jobParameters',
+    function numberOfParameters() {
```
Comment (adong) [inline app/components/pipeline-event-row/component.js:62]: I felt something here caused the display issue when pipeline is brand new and has no builds

**Review by adong — CHANGES_REQUESTED** [review]

**Comment by adong** [comment]
For _restart_ case:  If I restart on job p1, somehow main's job parameters still show up. I guess that's okay. Because `p1` could be an upstream job which triggers `main`

image

**Comment by sagar1312** [comment]
> ~Looks like validator works fine.~
> 
> Actually, it does not show individual jobs' parameters image

Changes have been pushed to show individual job parameters
image

File: app/components/pipeline-event-row/component.js (hunk 3 above)
Comment (sagar1312) [reply app/components/pipeline-event-row/component.js:62]: Done

**Comment by sagar1312** [comment]
> For brand new pipelines, only shared parameters shows up, job level parameters does not show. image
> 
> It shows after kick off the 1st build.

Fixed

**Review by adong — DISMISSED** [review]
Looks good to me. We can come back to visit this, if there's anything I overlooked,

**Review by adong — APPROVED** [review]

**Pull request merged**